scripts/network/network_manager.js

```diff
@@ -0,0 +1,192 @@
+import { ExplorerNetwork, Network, RPCNodeNetwork } from './network.js';
+import { cChainParams } from '../chain_params.js';
+import { fAutoSwitch } from '../settings.js';
+import { debugLog, DebugTopics, debugWarn } from '../debug.js';
+import { sleep } from '../utils.js';
+import { getEventEmitter } from '../event_bus.js';
+
+class NetWorkManager {
+    /**
+     * @type {Network} - Current selected Explorer
+     */
+    currentExplorer;
+
+    /**
+     * @type {Network} - Current selected RPC node
+     */
+    currentNode;
+
+    /**
+     * @type {Array<Network>} - List of all available Networks
+     */
+    networks = [];
+
+    start() {
+        this.networks = [];
+        for (let network of cChainParams.current.Explorers) {
+            this.networks.push(new ExplorerNetwork(network.url));
+        }
+        for (let network of cChainParams.current.Nodes) {
+            this.networks.push(new RPCNodeNetwork(network.url));
+        }
+    }
+
+    /**
+     * Reset the list of available networks
+     */
+    reset() {
+        this.networks = [];
+    }
+
+    /**
+     * Sets the network in use by MPW.
+     * @param {string} strUrl - network to use
+     * @param {boolean} isRPC - whether we are setting the explorer or the RPC node
+     */
+    setNetwork(strUrl, isRPC) {
+        if (this.networks.length === 0) {
+            this.start();
+        }
+        const found = this.networks.find(
+            (network) => network.strUrl === strUrl
+        );
+        if (!found) throw new Error('Cannot find provided Network!');
+        if (isRPC) {
+            this.currentNode = found;
+        } else {
+            this.currentExplorer = found;
+        }
+    }
+
+    /**
+     * Call all networks until one is succesful
+     * seamlessly attempt the same call on multiple other instances until success.
+     * @param {string} funcName - The function to re-attempt with
+     * @param {boolean} isRPC - Whether to begin with the selected explorer or RPC node
+     * @param  {...any} args - The arguments to pass to the function
+     */
+    async retryWrapper(funcName, isRPC, ...args) {
+        let nMaxTries = this.networks.length;
+        let attemptNet = isRPC
+            ? this.currentNode.copy()
+            : this.currentExplorer.copy();
+
+        let i = this.networks.findIndex((net) => attemptNet.equal(net));
+        if (i == -1) {
+            debugWarn(DebugTopics.NET, 'Cannot find index in networks array');
+            i = 0;
+        }
```

Comment:
Not sure why we need the whole `copy` `equal` thing. Can't we just hold the reference?

Reply:
you are right, I'm used to C++